Add support for the stack switching proposal

dhil opened the pull request:
This patch implements the [stack switching proposal](https://github.com/WebAssembly/stack-switching) (which at the time of writing is a [phase 2 proposal](https://github.com/WebAssembly/proposals)). 

The proposal adds a new composite type:
* `cont $ft` indexed by a function type, which is the type of a continuation.

Two new abstract continuation heap types and their subtyping hierarchy are added:
* `cont` the top type;
* and `nocont` the bottom type, i.e. `nocont <: cont`.

Six new instructions:
* `cont.new $ct`, parameterised by a continuation type index, which creates a new continuation.
* `cont.bind $ct-0 $ct-1`, parameterised by two continuation type indices, which partially applies a continuation of type `$ct-0`, resulting in a continuation of type `$ct-1`.
* `suspend $tag`, parameterised by a tag, which suspends the current context.
* `resume $ct (on $tag $label)*`, parameterised by a continuation type index and a jump table which maps tags to labels (similar to how `(catch $tag $label)` works in the exception handling proposal); this instruction resumes a continuation.
* `resume_throw $ct $tag (on $tag-1 $label)*`, parameterised by both a continuation type index and a tag index (in addition to the jump table). This instruction injects the exception tag `$tag` into the continuation argument.
* `switch $ct $tag`, parameterised by a continuation type index and a tag index. This instruction performs a direct context switch to a given continuation (i.e. a symmetric transfer of control).

The wellformedness condition for tags is relaxed, such that they are allowed to have a non-empty result type. This result type is used to validate the types of return values of `suspend $tag`.

The features of the proposal are gated under a new feature flag `stack-switching` (off by default). Most of the additions are test artifacts. The implementation of the proposal itself is modest.

<hr>

There are a few things in this patch that I am not quite sure about:

1. I have used a `UnpackedIndex` in the representation of the new indexed continuation type. In the validator I sometimes have to work with `CoreTypeId` and other times with raw `u32` indices. I am not convinced this is the best approach.
2. In validating `cont.bind` and the jump tables (aka resume tables) on `resume` and `resume_throw`, I have to construct and manipulate the underlying function types of continuations, in particular, I have to check subtyping relations between parts of these function types (c.f. `cont.bind` where I have to drop a prefix of the parameter list). So I have had to essentially reimplement function subtype checking. I have done this via a function `is_subtype_many` that applies the `is_subtype` function pointwise to two equally sized collections. I am not particularly fond of this approach.
3. I have not extended the `TagKind` structure. Though, an argument can be made for adding a further kind, lets call it `Control`, to differentiate exception and control tags internally. But I am not really sure whether the distinction is helpful.

alexcrichton commented on crates/wasmparser/src/validator/operators.rs:
```diff
@@ -1488,6 +1559,83 @@ where
+    /// (i.e. equally sized after skipped elements).
+    fn is_subtype_many(
+        &mut self,
+        skipcnt1: usize,
+        ts1: &[ValType],
+        skipcnt2: usize,
+        ts2: &[ValType],
```
Instead of taking a `skipcntN` could this only take two slices? It should be possible to chop off the front of the slices in callers explicitly if necessary

dhil replied:
Ah yes, it is much prettier. Thanks!

alexcrichton commented on crates/wasmparser/src/resources.rs:
```diff
@@ -50,6 +50,9 @@ pub trait WasmModuleResources {
     /// The sub type must be canonicalized.
     fn sub_type_at(&self, type_index: u32) -> Option<&SubType>;
 
+    /// Returns the `SubType` associated with the given core type id.
+    fn sub_type_at_id(&self, id: CoreTypeId) -> Option<&SubType>;
```
Unlike the other method in this trait which are called with possibly-arbitrary indices we know that `CoreTypeId` is always trusted so it's ok to return `&SubType` from this instead of `Option<_>`

alexcrichton commented on crates/wasmparser/src/validator/operators.rs:
```diff
@@ -1416,6 +1428,65 @@ where
         }
     }
 
+    fn cont_type_at(&self, key: TypeKey) -> Result<&ContType> {
+        let sub_ty = match key {
+            Either::A(at) => self.sub_type_at(at)?,
+            Either::B(id) => match self.resources.sub_type_at_id(id) {
+                None => bail!(self.offset, "unknown type {}", TypeIdentifier::index(&id)),
+                Some(sub_ty) => sub_ty,
+            },
+        };
+        if let CompositeInnerType::Cont(cont_ty) = &sub_ty.composite_type.inner {
+            if self.inner.shared && !sub_ty.composite_type.shared {
+                bail!(
+                    self.offset,
+                    "shared continuations cannot access unshared continuations",
+                );
+            }
+            Ok(cont_ty)
+        } else {
+            bail!(
+                self.offset,
+                "non-continuation type {}",
+                match key {
+                    Either::A(at) => at as usize,
+                    Either::B(id) => TypeIdentifier::index(&id),
+                }
+            )
+        }
+    }
```
I might recommend splitting this into `cont_type_at` which takes a `u32` and `const_type_at_id` which takes a `CoreTypeId`. I think that might be a bit easier to work with instead of working with `Either` perhaps? Functions like `func_type_of_cont_type_at` below could take `&ContType` instead of `TypeKey` in that case to let the caller call the appropriate `const_type_at*` function and then pass the result into later parts of validation.

alexcrichton commented on crates/wasmparser/src/validator/operators.rs:
```diff
@@ -1416,6 +1428,65 @@ where
+        let sub_ty = match key {
+            Either::A(at) => self.sub_type_at(at)?,
+            Either::B(id) => match self.resources.sub_type_at_id(id) {
+                None => bail!(self.offset, "unknown type {}", TypeIdentifier::index(&id)),
```
Here we won't want to surface `TypeIdentifier::index` since that's just a purely internal index during validation. In these cases it's ok to jus say "unknown type" or "non-continuation type" or something like that without mentioning the index.

alexcrichton commented on crates/wasmparser/src/validator/operators.rs:
```diff
@@ -1416,6 +1428,65 @@ where
+        }
+    }
+
+    fn func_type_of_cont_type_at(&self, key: TypeKey) -> Result<&'resources FuncType> {
+        let cont_ty = self.cont_type_at(key)?;
+        match cont_ty.0.as_core_type_id() {
+            None => bail!(self.offset, "invalid cont type"),
+            Some(func_id) => match self.resources.sub_type_at_id(func_id) {
+                None => bail!(
+                    self.offset,
+                    "unknown type {}",
+                    TypeIdentifier::index(&func_id)
+                ),
+                Some(sub_ty) => {
+                    if let CompositeInnerType::Func(func_ty) = &sub_ty.composite_type.inner {
+                        if self.inner.shared && !sub_ty.composite_type.shared {
+                            bail!(
+                                self.offset,
+                                "shared functions cannot access unshared functions",
+                            );
+                        }
+                        Ok(&func_ty)
+                    } else {
+                        bail!(
+                            self.offset,
+                            "non-function type {}",
+                            TypeIdentifier::index(&func_id)
+                        )
+                    }
+                }
+            },
+        }
+    }
```
This function, if it takes `&ContType`, I believe can be infallible (e.g. unwraps/expects/etc). The validation that `ContType` always points to a function should have already been validated when the type was added to the module, so here's it's safe to assume that invariant is always upheld.

dhil replied:
Great suggestion, thanks! After changing the index representation on `Cont` to `PackedIndex` everything else kind of falls into place. It is much nicer.

alexcrichton reviewed (commented):
Thanks so much again for working on this here and being willing to rebase your work while it was in development, it's much appreciated! I've left an initial round of comments below but I'll want to dig in more later to some of your concerns around duplication of function type subtyping as well after they're handled. Feel free to push back on anything I'm saying below as well that doesn't make sense too

dhil commented on crates/wasmparser/src/readers/core/operators.rs:
```diff
@@ -477,3 +477,54 @@ impl<'a> FromReader<'a> for Catch {
+            0x01 => Handle::OnSwitch {
+                tag: reader.read_var_u32()?,
+            },
+            x => return reader.invalid_leading_byte(x, "on"),
```
How about "on clause"?

dhil commented on crates/wasm-encoder/src/core/types.rs:
```diff
@@ -611,6 +625,13 @@ impl<'a> CoreTypeEncoder<'a> {
 
+    fn encode_cont(&mut self, ty: &ContType) {
+        self.bytes.push(0x5d);
+        // Note that this is encoded as a signed type rather than unsigned
+        // as it's decoded as an s33
+        i64::from(ty.0).encode(self.bytes);
```
I think, I copy-pasted it from my downstream implementation, where I think I copy pasted it from how `HeapType::Concrete` is encoded. Actually, I think I modelled a lot of `Cont` on how `Concrete` works for heap types. But maybe it fine to encode and decode continuation types as `u32`s?

dhil replied:
Actually, I retract the above. I looked at the reference implementation. There we encode it as a `s33`. But we have written in the explainer/informal spec that it is supposed to be encoded as a `typeidx`, which I believe per the spec should be encoded as an `u32`. So, there seems to be an inconsistency here. I will chat with @rossberg about it.

alexcrichton replied:
I believe `s33` is used in binary encoding locations where a positive value indicates "this is a type index" and a negative value indicates "this is actually a construct which isn't a type index but it's a structure to parse". It's needed to make the binary representation more compact by not needing a prefix byte for differentiating between the two. In the context of `Cont`, though, I don't think it's necessary because the prefix byte of `Cont`, 0x5d, has already been pushed so there's no ambiguity of what the next item is. That's what makes me think that this may want to be a u32 instead of an s33, but I wanted to confirm to be sure

dhil replied:
I think you are correct. I spoke with Andreas about this. We decided to go with the `s33` encoding as it may be more forwards compatible as it gives us some extension point for adding future attributes to continuation types (should we ever want that). So I have proposed this change on the spec repo https://github.com/WebAssembly/stack-switching/pull/89.

dhil commented on crates/wasmparser/src/validator/core/canonical.rs:
```diff
@@ -246,6 +246,35 @@ pub(crate) trait InternRecGroup {
+                match types[id].composite_type.inner {
+                    CompositeInnerType::Func(_) => (),
+                    _ => {
+                        return Err(BinaryReaderError::new(
+                            crate::validator::format!("non-function type {}", id.index()),
+                            offset,
+                        ))
```
Yes, absolutely! I'll fix that.

dhil commented on crates/wasmparser/src/validator/core/canonical.rs:
```diff
@@ -246,6 +246,35 @@ pub(crate) trait InternRecGroup {
+                }
+                // Check that the type index points to a valid function type.
+                let id = match t.0 {
+                    UnpackedIndex::Id(id) => id,
+                    UnpackedIndex::Module(idx) => self.type_id_at(idx, offset)?,
```
Good point! Thanks!

dhil commented on crates/wasmparser/src/validator/core.rs:
```diff
@@ -946,7 +946,7 @@ impl Module {
         }
         let ty = self.func_type_at(ty.func_type_idx, types, offset)?;
-        if !ty.results().is_empty() {
+        if !ty.results().is_empty() && !features.stack_switching() {
```
Yes, good point! That's an oversight from me.

dhil commented on crates/wasmprinter/src/operator.rs:
```diff
@@ -667,6 +691,35 @@ macro_rules! define_visit {
         $self.push_str(" ")?;
         $self.printer.print_field_idx($self.state, $ty, $field)?;
     );
+    (payload $self:ident ContNew $type_index:ident) => (
+        $self.push_str(" ")?;
+        $self.printer.print_idx(&$self.state.core.type_names, $type_index)?;
+    );
+    (payload $self:ident ContBind $argument_index:ident $result_index:ident) => (
+        $self.push_str(" ")?;
+        $self.printer.print_idx(&$self.state.core.type_names, $argument_index)?;
+        $self.push_str(" ")?;
+        $self.printer.print_idx(&$self.state.core.type_names, $result_index)?;
+    );
+    (payload $self:ident Suspend $tag_index:ident) => (
+        $self.tag_index($tag_index)?;
+    );
+    (payload $self:ident Resume $type_index:ident $table:ident) => (
+        $self.push_str(" ")?;
+        $self.printer.print_idx(&$self.state.core.type_names, $type_index)?;
+        $self.resume_table($table)?;
+    );
+    (payload $self:ident ResumeThrow $type_index:ident $tag_index:ident $table:ident) => (
+        $self.push_str(" ")?;
+        $self.printer.print_idx(&$self.state.core.type_names, $type_index)?;
+        $self.tag_index($tag_index)?;
+        $self.resume_table($table)?;
+    );
+    (payload $self:ident Switch $type_index:ident $tag_index:ident) => (
+        $self.push_str(" ")?;
+        $self.printer.print_idx(&$self.state.core.type_names, $type_index)?;
+        $self.tag_index($tag_index)?;
+    );
```
Yes, I'll fix that.

dhil commented on crates/wasmparser/src/readers/core/types.rs:
```diff
@@ -776,6 +797,27 @@ pub struct StructType {
 
+/// Represents a type of a continuation in a WebAssembly module.
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct ContType(pub UnpackedIndex);
```
Excellent suggestion! I like this much better. Thanks!

dhil commented:
Thanks for your feedback! I think the key change was to represent the index on continuation types as `PackedIndex`. It makes everything much nicer.

alexcrichton commented on crates/wasmparser/src/readers/core/types.rs:
```diff
@@ -776,6 +797,24 @@ pub struct StructType {
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct ContType(pub PackedIndex);
+
+impl ContType {
+    /// Maps any `UnpackedIndex` via the specified closure.
+    #[cfg(feature = "validate")]
+    pub(crate) fn remap_indices(
+        &mut self,
+        map: &mut dyn FnMut(&mut PackedIndex) -> Result<()>,
+    ) -> Result<()> {
+        let mut idx = self.0;
+        map(&mut idx)?;
+        *self = ContType(idx);
+        Ok(())
+    }
+}
```
If you'd like it's probably ok to omit this entirely and replace the caller with `map(&mut ty.0)?;`

alexcrichton commented on crates/wasmparser/src/validator/core/canonical.rs:
```diff
@@ -246,6 +246,29 @@ pub(crate) trait InternRecGroup {
+                    );
+                }
+                // Check that the type index points to a valid function type.
+                let id = match t.0.as_core_type_id() {
+                    None => bail!(offset, "invalid continuation type index"),
+                    Some(id) => id,
+                };
```
Here it's ok to `.unwrap()` the result since a failure here means a bug in wasmparser itself rather than something the user can fix with the input wasm module

alexcrichton commented on crates/wasmparser/src/validator/operators.rs:
```diff
@@ -1416,12 +1416,46 @@ where
+        }
+    }
+
+    fn func_type_of_cont_type(&self, cont_ty: &ContType) -> Result<&'resources FuncType> {
+        let func_id = cont_ty.0.as_core_type_id().expect("valid core type id");
+        Ok(self.resources.sub_type_at_id(func_id).unwrap_func())
+    }
```
This looks like it should be possible to remove the `Result` return value and just return `&'resources FuncType` here

alexcrichton commented on crates/wasmparser/src/readers/core/types/matches.rs:
```diff
@@ -252,6 +259,40 @@ impl Matches for WithRecGroup<StorageType> {
+        match (*a, *b) {
+            (ContType(ca), ContType(cb)) => {
+                ca == cb || {
+                    let core_id_a = if let Some(ida) = ca.as_core_type_id() {
+                        ida
+                    } else {
+                        types
+                            .at_canonicalized_unpacked_index(
+                                WithRecGroup::rec_group(a),
+                                ca.unpack(),
+                                usize::MAX,
+                            )
+                            .expect("type references are checked during canonicalization")
+                    };
```
This looks like it may be a duplicate of `let core_type_id = |...| ...;` in validator/types.rs, would you be up for extracting that to a helper function on `TypeList` to help share the implementation between these locations?

alexcrichton replied:
Alternatively what you could also perhaps do here is to manufacture a `RefType`  and then call `reftype_is_subtype_impl` perhaps?

dhil replied:
Yes! I opted for the second approach.

alexcrichton commented on crates/wast/src/core/binary.rs:
```diff
@@ -342,6 +342,18 @@ impl From<&ArrayType<'_>> for wasm_encoder::ArrayType {
     }
 }
 
+impl Encode for ContType<'_> {
+    fn encode(&self, e: &mut Vec<u8>) {
+        self.0.encode(e);
+    }
+}
```
If you'd like it should be ok to remove this now as it's no longer needed (just a minor thing though)

alexcrichton reviewed (commented):
Ok a few small comments here and there, but otherwise this looks good to go to me 👍 . Thanks again!

alexcrichton commented:
And no rush of course! Just wanted to check in

dhil commented:
Thanks Alex! I have implemented your feedback now. The issue of whether we will encode the index on continuation types as an `u32` or a `s33` is yet to be resolved on the spec repo. In the meantime I have reverted back to the `s33` encoding to stay in sync with the reference interpreter.

dhil commented:
Thanks for your help!
